packages/vitest/src/integrations/coverage.ts

```diff
@@ -36,47 +39,79 @@ export function resolveC8Options(options: C8Options, root: string): ResolvedC8Op
 
   resolved.reporter = toArray(resolved.reporter)
   resolved.reportsDirectory = resolve(root, resolved.reportsDirectory)
+  resolved.tempDirectory = process.env.NODE_V8_COVERAGE || resolve(resolved.reportsDirectory, 'tmp')
 
   return resolved as ResolvedC8Options
 }
 
 export async function cleanCoverage(options: ResolvedC8Options, clean = true) {
   if (clean && existsSync(options.reportsDirectory))
     await fs.rm(options.reportsDirectory, { recursive: true, force: true })
+
+  if (!existsSync(options.tempDirectory))
+    await fs.mkdir(options.tempDirectory, { recursive: true })
 }
 
 const require = createRequire(import.meta.url)
 
 export async function reportCoverage(ctx: Vitest) {
+  // Flush coverage to disk
+  takeCoverage()
+
   // eslint-disable-next-line @typescript-eslint/no-var-requires
   const createReport = require('c8/lib/report')
   const report = createReport(ctx.config.coverage)
 
-  report._loadReports = () => ctx.coverage
-
-  const original = report._getMergedProcessCov
-
-  report._getMergedProcessCov = () => {
-    const r = original.call(report)
-
-    // add source maps
-    Array
-      .from(ctx.vitenode.fetchCache.entries())
-      .filter(i => !i[0].includes('/node_modules/'))
-      .forEach(([file, { result }]) => {
-        const map = result.map
-        if (!map)
-          return
-        const url = pathToFileURL(file).href
-        const sources = map.sources.length
-          ? map.sources.map(i => pathToFileURL(i).href)
-          : [url]
-        report.sourceMapCache[url] = {
-          data: { ...map, sources },
-        }
-      })
-
-    return r
+  // add source maps
+  const sourceMapMata: Record<string, { map: RawSourceMap; source: string | undefined }> = {}
+  await Promise.all(Array
+    .from(ctx.vitenode.fetchCache.entries())
+    .filter(i => !i[0].includes('/node_modules/'))
+    .map(async([file, { result }]) => {
+      const map = result.map
+      if (!map)
+        return
+
+      const url = pathToFileURL(file).href
+
+      let code: string | undefined
+      try {
+        code = (await fs.readFile(file)).toString()
+      }
+      catch {}
+
+      const sources = map.sources.length
+        ? map.sources.map(i => pathToFileURL(i).href)
+        : [url]
+
+      sourceMapMata[url] = {
+        source: result.code,
+        map: {
+          sourcesContent: code ? [code] : undefined,
+          ...map,
+          sources,
+        },
+      }
+    }))
+
+  // This is a magic number. It corresponds to the amount of code
+  // that we add in packages/vite-node/src/client.ts:114 (vm.runInThisContext)
+  // TODO: Include our transformations in soucemaps
+  const offset = 190
```

Comment:
Later we should calculate this number based on the length of the arguments, but I am fine to have the magic number for now.